src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
+{
+    /// <summary>
+    /// Caches the semantic token information that needs to be preserved between multiple
+    /// semantic token requests.
+    /// Multiple token sets can be cached per document. The number of token sets cached
+    /// per document is determined by the _maxCachesPerDoc field.
+    /// </summary>
+    [Export(typeof(SemanticTokensCache)), Shared]
+    internal class SemanticTokensCache
+    {
+        /// <summary>
+        /// Multiple cache requests or updates may be received concurrently.
+        /// We need this sempahore to ensure that we aren't making concurrent
+        /// modifications to the Tokens or NextResultId dictionaries.
+        /// </summary>
+        private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        /// <summary>
+        /// Number of cached token sets we store per document. Must be >= 1.
+        /// </summary>
+        private readonly int _maxCachesPerDoc = 5;
+
+        /// <summary>
+        /// The next resultId available to use.
+        /// </summary>
+        private long _nextResultId;
+
+        /// <summary>
+        /// Maps a document URI to its n most recently cached token sets.
+        /// </summary>
+        private Dictionary<Uri, List<LSP.SemanticTokens>> Tokens { get; }
```

Comment:
One thing to note about the cache being a MEF part is that these cached semantic tokens will be kept alive if the user closes the Roslyn workspace and opens a different project. Probably not a big deal, since this will eventually be OOP, but worth noting.

Reply:
readonly?